File: web-comics-viewer/src/comics-viewer.js
/**
 * Comics Viewer - Main Entry Point
 * Render engine for interactive .comics archives with scroll-driven animations
 */

import { parseComicsData } from './models.js';
import { AnimationProcessor } from './animation.js';
import { TileLoader } from './tile-loader.js';
import { SoundManager } from './sound-manager.js';

export class ComicsViewer {
  static TILE_SIZE = 512;

  constructor(container, options = {}) {
    this.container = container;
    this.options = {
      languageIndex: 0,
      soundEnabled: true,
      onLoad: null,
      onScroll: null,
      onError: null,
      onLayerTap: null,
      ...options
    };

    this.comics = null;
    this.archive = null;
    this.animationProcessor = null;
    this.tileLoader = null;
    this.soundManager = null;
    this.contentElement = null;
    this.layerElements = [];
    this.currentScroll = 0;

    this._setupContainer();
    this._bindEvents();
  }

  _setupContainer() {
    this.container.classList.add('comics-viewer');
    this.contentElement = document.createElement('div');
    this.contentElement.className = 'comics-content';
    this.container.appendChild(this.contentElement);
  }

  _bindEvents() {
    this.container.addEventListener('scroll', this._onScroll.bind(this), { passive: true });
  }

  async load(archiveUrl) {
    try {
      // Dynamic import of fflate for ZIP parsing
      const { unzipSync } = await import('https://cdn.jsdelivr.net/npm/fflate@0.8.2/+esm');

      const response = await fetch(archiveUrl);
      if (!response.ok) {
        throw new Error(`Failed to fetch: ${response.status}`);
      }

      const buffer = await response.arrayBuffer();
      this.archive = unzipSync(new Uint8Array(buffer));

      // Parse data.json
      const dataJson = this.archive['data.json'];
      if (!dataJson) {
        throw new Error('data.json not found in archive');
      }

      const jsonString = new TextDecoder().decode(dataJson);
      this.comics = parseComicsData(JSON.parse(jsonString));

      // Initialize subsystems
      this.animationProcessor = new AnimationProcessor(this.comics);
      this.tileLoader = new TileLoader(this.archive, this.comics, this.options.languageIndex);
      this.soundManager = new SoundManager(this.archive, this.comics, this.options.soundEnabled);

      // Setup content
      this._setupContent();

      // Notify loaded
      if (this.options.onLoad) {
        this.options.onLoad({
          width: this.comics.width,
          height: this.comics.height,
          layerCount: this.comics.layers.length,
          hasSound: this.comics.sounds && this.comics.sounds.length > 0
        });
      }

      // Initial render
      this._onScroll();

    } catch (error) {
      console.error('ComicsViewer load error:', error);
      if (this.options.onError) {
        this.options.onError(error.message);
      }
      throw error;
    }
  }

  _setupContent() {
    const { width, height, layers } = this.comics;

    // Set content size
    this.contentElement.style.width = `${width}px`;
    this.contentElement.style.height = `${height}px`;

    // Create layer elements (bottom to top)
    layers.forEach((layer, index) => {
      const layerEl = document.createElement('div');
      layerEl.className = 'comics-layer';
      layerEl.dataset.layerId = layer.id || index;
      layerEl.style.width = `${layer.width}px`;
      layerEl.style.height = `${layer.height}px`;
      layerEl.style.left = `${layer.x}px`;
      layerEl.style.top = `${layer.y}px`;

      // Create tile grid
      this.tileLoader.createTileElements(layerEl, layer, index);

      this.contentElement.appendChild(layerEl);
      this.layerElements.push(layerEl);
    });
  }

  _onScroll() {
    const scrollY = this.container.scrollTop;
    const maxScroll = this.contentElement.offsetHeight - this.container.offsetHeight;
    this.currentScroll = scrollY;

    // Process animations
    if (this.animationProcessor) {
      const transforms = this.animationProcessor.process(scrollY);
      this._applyTransforms(transforms);
    }

    // Update tiles visibility
    if (this.tileLoader) {
      this.tileLoader.updateViewport(scrollY, this.container.offsetHeight);
    }

    // Process sounds
    if (this.soundManager) {
      this.soundManager.process(scrollY);
    }

    // Notify callback
    if (this.options.onScroll) {
      this.options.onScroll(scrollY, maxScroll);
    }
  }

  _applyTransforms(transforms) {
    transforms.forEach((transform, index) => {
      const layerEl = this.layerElements[index];
      if (!layerEl) return;

      const { translateX, translateY, scaleX, scaleY, rotation, alpha } = transform;

      // Build transform string: Scale -> Rotate -> Translate
      const transformStr = `translate(${translateX}px, ${translateY}px) rotate(${rotation}deg) scale(${scaleX}, ${scaleY})`;

      layerEl.style.transform = transformStr;
      layerEl.style.opacity = alpha;
    });
  }

  setScrollOffset(offset) {
    this.container.scrollTop = offset;
  }

  getScrollOffset() {
    return this.container.scrollTop;
  }

  setLanguageIndex(index) {
    this.options.languageIndex = index;
    if (this.tileLoader) {
      this.tileLoader.setLanguageIndex(index);
    }
  }

  setSoundEnabled(enabled) {
    this.options.soundEnabled = enabled;
    if (this.soundManager) {
      this.soundManager.setEnabled(enabled);
    }
  }

  pauseSounds() {
    this.soundManager?.pause();
  }

  resumeSounds() {
    this.soundManager?.resume();
  }

  destroy() {
    this.soundManager?.destroy();
    this.tileLoader?.destroy();
    this.container.innerHTML = '';
    this.comics = null;
    this.archive = null;
  }
}

export default ComicsViewer;

File: engines/unity_comics.engine/Runtime/ComicsViewer.cs
using System;
using UnityEngine;
using NativeMind.ComicsViewer.Models;
using NativeMind.ComicsViewer.Core;
using NativeMind.ComicsViewer.IO;
using NativeMind.ComicsViewer.Rendering;
using NativeMind.ComicsViewer.Audio;

// Backwards compatibility alias
using ZipArchiveProvider = NativeMind.ComicsViewer.IO.ZipArchiveSource;

namespace NativeMind.ComicsViewer
{
    /// <summary>
    /// Information about loaded comics
    /// </summary>
    public struct ComicsInfo
    {
        public int Width;
        public int Height;
        public int LayerCount;
        public bool HasSound;
    }

    /// <summary>
    /// Main comics viewer component
    /// Attach to a GameObject with Camera child for WorldSpace rendering
    /// </summary>
    public class ComicsViewer : MonoBehaviour
    {
        [Header("Configuration")]
        [Tooltip("Path to .comics archive file")]
        public string archivePath;

        [Tooltip("Language index for localized content")]
        [Range(0, 10)]
        public int languageIndex = 0;

        [Tooltip("Enable sound playback")]
        public bool soundEnabled = true;

        [Header("Scroll Settings")]
        [Tooltip("Scroll speed multiplier")]
        public float scrollSpeed = 1f;

        [Tooltip("Enable scroll momentum")]
        public bool enableMomentum = true;

        [Tooltip("Momentum friction")]
        [Range(0.9f, 0.99f)]
        public float momentumFriction = 0.95f;

        // Events
        public event Action<ComicsInfo> OnLoaded;
        public event Action<float, float> OnScroll; // scrollY, maxScroll
        public event Action<string> OnError;

        // Private state
        private Comics _comics;
        private IComicsSource _source;
        private AnimationProcessor _animationProcessor;
        private TileRenderer _tileRenderer;
        private SoundManager _soundManager;

        private Transform _contentRoot;
        private Camera _viewCamera;

        private float _scrollY;
        private float _maxScroll;
        private float _velocity;
        private bool _isDragging;
        private float _lastDragY;
        private float _viewportHeight;

        public float ScrollOffset => _scrollY;
        public float MaxScrollOffset => _maxScroll;
        public bool IsLoaded => _comics != null;

        private void Awake()
        {
            // Create content root
            _contentRoot = new GameObject("ComicsContent").transform;
            _contentRoot.SetParent(transform, false);

            // Find or create camera
            _viewCamera = GetComponentInChildren<Camera>();
            if (_viewCamera == null)
            {
                var camObj = new GameObject("ViewCamera");
                camObj.transform.SetParent(transform, false);
                _viewCamera = camObj.AddComponent<Camera>();
                _viewCamera.orthographic = true;
                _viewCamera.clearFlags = CameraClearFlags.SolidColor;
                _viewCamera.backgroundColor = Color.black;
            }
        }

        private void Start()
        {
            if (!string.IsNullOrEmpty(archivePath))
            {
                LoadArchive(archivePath);
            }
        }

        /// <summary>
        /// Load comics from archive path
        /// </summary>
        public void LoadArchive(string path)
        {
            archivePath = path;
            var source = new ZipArchiveSource(path);
            source.Prepare();
            Initialize(source);
        }

        /// <summary>
        /// Load comics from unpacked folder (for editor preview)
        /// </summary>
        public void LoadFolder(string folderPath)
        {
            var source = new FolderSource(folderPath);
            source.Prepare();
            Initialize(source);
        }

        /// <summary>
        /// Initialize with any source (for dependency injection)
        /// </summary>
        public void Initialize(IComicsSource source)
        {
            try
            {
                // Cleanup previous
                Unload();

                _source = source;

                // Parse data.json
                string jsonString = _source.ReadDataJson();
                _comics = ComicsParser.Parse(jsonString);

                // Setup camera for content size
                SetupCamera();

                // Initialize systems
                _animationProcessor = new AnimationProcessor(_comics);
                _tileRenderer = new TileRenderer(_comics, _source, _contentRoot, languageIndex);

                // Sound manager only for read-only sources (runtime)
                // Editor can enable separately if needed
                if (_source.IsReadOnly)
                {
                    _soundManager = new SoundManager(_source, _comics, soundEnabled);
                }

                // Calculate scroll bounds
                _maxScroll = Mathf.Max(0, _comics.height - _viewportHeight);

                // Initial render
                UpdateViewport();

                // Notify loaded
                OnLoaded?.Invoke(new ComicsInfo
                {
                    Width = _comics.width,
                    Height = _comics.height,
                    LayerCount = _comics.layers.Count,
                    HasSound = _comics.sounds?.Count > 0
                });
            }
            catch (Exception e)
            {
                Debug.LogError($"Failed to load comics: {e.Message}");
                OnError?.Invoke(e.Message);
            }
        }

        /// <summary>
        /// Refresh from source after editor modifies document
        /// </summary>
        public void RefreshFromSource()
        {
            if (_source == null) return;

            try
            {
                _source.Invalidate();

                // Reload data
                string jsonString = _source.ReadDataJson();
                _comics = ComicsParser.Parse(jsonString);

                // Recreate animation processor
                _animationProcessor = new AnimationProcessor(_comics);

                // Invalidate tile cache
                _tileRenderer?.InvalidateCache();

                // Recalculate bounds
                _maxScroll = Mathf.Max(0, _comics.height - _viewportHeight);

                // Re-render
                UpdateViewport();
            }
            catch (Exception e)
            {
                Debug.LogError($"Failed to refresh comics: {e.Message}");
                OnError?.Invoke(e.Message);
            }
        }

        private void SetupCamera()
        {
            // Position camera at content center
            _viewCamera.transform.localPosition = new Vector3(
                _comics.width / 2f,
                0,
                -10
            );

            // Calculate orthographic size based on screen
            float screenAspect = (float)Screen.width / Screen.height;
            float comicsAspect = (float)_comics.width / _comics.height;

            if (screenAspect > comicsAspect)
            {
                // Screen is wider - fit to height
                _viewCamera.orthographicSize = Screen.height / 2f;
            }
            else
            {
                // Screen is taller - fit to width
                _viewCamera.orthographicSize = _comics.width / (2f * screenAspect);
            }

            _viewportHeight = _viewCamera.orthographicSize * 2f;
        }

        private void Update()
        {
            if (_comics == null) return;

            HandleInput();
            UpdateMomentum();
            UpdateViewport();
        }

        private void HandleInput()
        {
            // Mouse/touch drag
            if (Input.GetMouseButtonDown(0))
            {
                _isDragging = true;
                _lastDragY = Input.mousePosition.y;
                _velocity = 0;
            }
            else if (Input.GetMouseButtonUp(0))
            {
                _isDragging = false;
            }

            if (_isDragging)
            {
                float deltaY = _lastDragY - Input.mousePosition.y;
                float scrollDelta = deltaY * scrollSpeed;

                _velocity = scrollDelta;
                SetScrollOffset(_scrollY + scrollDelta);

                _lastDragY = Input.mousePosition.y;
            }

            // Mouse wheel
            float scroll = Input.mouseScrollDelta.y;
            if (Mathf.Abs(scroll) > 0.01f)
            {
                SetScrollOffset(_scrollY - scroll * 100 * scrollSpeed);
            }
        }

        private void UpdateMomentum()
        {
            if (!enableMomentum || _isDragging) return;

            if (Mathf.Abs(_velocity) > 0.1f)
            {
                SetScrollOffset(_scrollY + _velocity);
                _velocity *= momentumFriction;
            }
            else
            {
                _velocity = 0;
            }
        }

        private void UpdateViewport()
        {
            if (_tileRenderer == null) return;

            // Update camera position for scroll
            var camPos = _viewCamera.transform.localPosition;
            camPos.y = -_scrollY - _viewportHeight / 2f;
            _viewCamera.transform.localPosition = camPos;

            // Update tile visibility
            _tileRenderer.UpdateViewport(_scrollY, _viewportHeight);

            // Process animations
            var transforms = _animationProcessor.Process(_scrollY);
            _tileRenderer.ApplyTransforms(transforms);

            // Process sounds
            _soundManager?.Process(_scrollY);
        }

        /// <summary>
        /// Set scroll offset programmatically
        /// </summary>
        public void SetScrollOffset(float offset)
        {
            float newScroll = Mathf.Clamp(offset, 0, _maxScroll);

            if (Mathf.Abs(newScroll - _scrollY) > 0.01f)
            {
                _scrollY = newScroll;
                OnScroll?.Invoke(_scrollY, _maxScroll);
            }
        }

        /// <summary>
        /// Set language index for localized content
        /// </summary>
        public void SetLanguageIndex(int index)
        {
            languageIndex = index;
            _tileRenderer?.SetLanguageIndex(index);
        }

        /// <summary>
        /// Enable or disable sound
        /// </summary>
        public void SetSoundEnabled(bool enabled)
        {
            soundEnabled = enabled;
            _soundManager?.SetEnabled(enabled);
        }

        /// <summary>
        /// Pause all sounds
        /// </summary>
        public void PauseSounds()
        {
            _soundManager?.Pause();
        }

        /// <summary>
        /// Resume sounds
        /// </summary>
        public void ResumeSounds()
        {
            _soundManager?.Resume();
        }

        /// <summary>
        /// Unload current comics
        /// </summary>
        public void Unload()
        {
            _soundManager?.Dispose();
            _soundManager = null;

            _tileRenderer?.Dispose();
            _tileRenderer = null;

            _animationProcessor = null;

            _source?.Dispose();
            _source = null;

            _comics = null;
            _scrollY = 0;
            _maxScroll = 0;
        }

        /// <summary>
        /// Get the current comics source (for advanced usage)
        /// </summary>
        public IComicsSource Source => _source;

        private void OnDestroy()
        {
            Unload();
        }

        private void OnApplicationPause(bool pause)
        {
            if (pause)
            {
                PauseSounds();
            }
            else
            {
                ResumeSounds();
            }
        }
    }
}
